internal/attestation/azure/snp/validator.go

```diff
@@ -52,235 +55,154 @@ func NewValidator(cfg *config.AzureSEVSNP, log attestation.Logger) *Validator {
 		maa:          newMAAClient(),
 		config:       cfg,
 		log:          log,
+		getter:       trust.DefaultHTTPSGetter(),
 	}
 	v.Validator = vtpm.NewValidator(
 		cfg.Measurements,
 		v.getTrustedKey,
-		validateCVM,
+		// stub, since SEV-SNP attestation is already verified in trustedKeyFromSNP().
+		func(vtpm.AttestationDocument, *attest.MachineState) error {
+			return nil
+		},
 		log,
 	)
 	return v
 }
 
-// validateCVM is a stub, since SEV-SNP attestation is already verified in trustedKeyFromSNP().
-func validateCVM(vtpm.AttestationDocument, *attest.MachineState) error {
-	return nil
-}
-
-func newSNPReportFromBytes(reportRaw []byte) (snpAttestationReport, error) {
-	var report snpAttestationReport
-	if err := binary.Read(bytes.NewReader(reportRaw), binary.LittleEndian, &report); err != nil {
-		return snpAttestationReport{}, fmt.Errorf("reading attestation report: %w", err)
-	}
-
-	return report, nil
-}
-
-func reverseEndian(b []byte) {
-	for i := 0; i < len(b)/2; i++ {
-		b[i], b[len(b)-i-1] = b[len(b)-i-1], b[i]
-	}
-}
-
 // getTrustedKey establishes trust in the given public key.
-// It does so by verifying the SNP attestation statement in instanceInfo.
+// It does so by verifying the SNP attestation document.
 func (v *Validator) getTrustedKey(ctx context.Context, attDoc vtpm.AttestationDocument, extraData []byte) (crypto.PublicKey, error) {
+	// transform the instanceInfo received from Microsoft into a verifiable attestation report format.
 	var instanceInfo azureInstanceInfo
 	if err := json.Unmarshal(attDoc.InstanceInfo, &instanceInfo); err != nil {
-		return nil, fmt.Errorf("unmarshalling instanceInfoRaw: %w", err)
+		return nil, fmt.Errorf("unmarshalling instanceInfo: %w", err)
 	}
-
-	report, err := newSNPReportFromBytes(instanceInfo.AttestationReport)
+	att, err := instanceInfo.attestationWithCerts(v.getter)
 	if err != nil {
 		return nil, fmt.Errorf("parsing attestation report: %w", err)
 	}
 
-	vcek, err := v.validateVCEK(instanceInfo.Vcek, instanceInfo.CertChain)
-	if err != nil {
-		return nil, fmt.Errorf("validating VCEK: %w", err)
-	}
-
-	if err := v.validateSNPReport(ctx, vcek, report, instanceInfo.MAAToken, extraData); err != nil {
-		return nil, fmt.Errorf("validating SNP report: %w", err)
-	}
-
+	// Verify the attestation report's certificates.
+	if err := verify.SnpAttestation(att, &verify.Options{}); err != nil {
+		return nil, fmt.Errorf("verifying SNP attestation: %w", err)
+	}
+
+	// Checks if the attestation report matches the given constraints.
+	// Some constraints are implicitly checked by validate.SnpAttestation:
+	// - the report is not expired
+	if err := validate.SnpAttestation(att, &validate.Options{
```

Comment:
We should check that `VMPL`  is 0.
This is a check that was missing in our old implementation.